plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs

```diff
@@ -0,0 +1,251 @@
+cabalRules recorder = do
+  -- Make sure we initialise the cabal files-of-interest.
+  ofInterestRules recorder
+  -- Rule to produce diagnostics for cabal files.
+  define (cmapWithPrio LogShake recorder) $ \ParseCabal file -> do
+    t <- use GetModificationTime file
+    log' Debug $ LogModificationTime file t
+    mVirtualFile <- Shake.getVirtualFile file
+    contents <- case mVirtualFile of
+      Just vfile -> pure $ Encoding.encodeUtf8 $ VFS.virtualFileText vfile
+      Nothing -> do
+        liftIO $ BS.readFile $ fromNormalizedFilePath file
+
+    (pWarnings, pm) <- liftIO $ Parse.parseCabalFileContents contents
+    let warningDiags = fmap (Diagnostics.warningDiagnostic file) pWarnings
+    case pm of
+      Left (_cabalVersion, pErrorNE) -> do
+        let errorDiags = NE.toList $ NE.map (Diagnostics.errorDiagnostic file) pErrorNE
+            allDiags = errorDiags <> warningDiags
+        log' Debug $ LogDiagnostics file allDiags
+        pure (allDiags, Nothing)
+      Right _ -> do
+        log' Debug $ LogDiagnostics file warningDiags
+        pure (warningDiags, Just ())
+
+  action $ do
+    -- Run the cabal kick. This code always runs when 'shakeRestart' is run.
+    -- Must be careful to not impede the performance too much. Crucial to
+    -- a snappy IDE experience.
+    kick
+  where
+    log' = logWith recorder
+
+-- | This is the kick function for the cabal plugin.
+-- We run this action, whenever we need to restart the shake session, which triggers
+-- actions to produce diagnostics for cabal files.
+--
+-- It is paramount that this kick-function can be run quickly, since it is a blocking
+-- function invocation.
+kick :: Action ()
+kick = do
+  files <- HashMap.keys <$> getCabalFilesOfInterestUntracked
+  void $ uses ParseCabal files
+
+-- ----------------------------------------------------------------
+-- Code Actions
+-- ----------------------------------------------------------------
+
+licenseSuggestCodeAction
+  :: IdeState
+  -> PluginId
+  -> CodeActionParams
+  -> LspM Config (Either ResponseError (ResponseResult 'TextDocumentCodeAction))
+licenseSuggestCodeAction _ _ (CodeActionParams _ _ (TextDocumentIdentifier uri) _range CodeActionContext{_diagnostics=List diags}) =
+  pure $ Right $ List $ mapMaybe (fmap InR . LicenseSuggest.licenseErrorAction uri) diags
+
+-- ----------------------------------------------------------------
+-- Cabal file of Interest rules and global variable
+-- ----------------------------------------------------------------
+
+-- | Cabal files that are currently open in the lsp-client.
+-- Specific actions happen when these files are saved, closed or modified,
+-- such as generating diagnostics, re-parsing, etc...
+--
+-- We need to store the open files to parse them again if we restart the shake session.
+-- Restarting of the shake session happens whenever these files are modified.
+newtype OfInterestCabalVar = OfInterestCabalVar (Var (HashMap NormalizedFilePath FileOfInterestStatus))
```

Comment:
I get that we do this for Haskell too... but I wonder why we can't e.g. just traverse the VFS and look for files matching the extension we care about?

Reply:
having a separate variable seemed like a better seperation to me. Otherwise, you need to remember to always call the correct function...